File: CustomApplications/VB.Net/GraphicsHowTo/Picking/PickRectangularCodeSnippet.vb
Imports System.IO
Imports System.Collections.Generic

#Region "UsingDirectives"

Imports System.Collections.ObjectModel
Imports System.Drawing
Imports AGI.STKGraphics
Imports AGI.STKObjects
Imports AGI.STKUtil
Imports AGI.STKVgt

#End Region

Namespace Picking
	Public Class PickRectangularCodeSnippet
		Inherits CodeSnippet
		Public Sub New()
            MyBase.New("Picking\PickRectangularCodeSnippet.vb")
		End Sub

		Public Sub MouseMove(scene As IAgStkGraphicsScene, root As AgStkObjectRoot, mouseX As Integer, mouseY As Integer)
            Dim manager2 As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager
            Dim overlayManager As IAgStkGraphicsScreenOverlayCollectionBase = DirectCast(manager2.ScreenOverlays.Overlays, IAgStkGraphicsScreenOverlayCollectionBase)
            Dim SelectedModels As List(Of IAgStkGraphicsModelPrimitive) = m_SelectedModels
			If m_Models IsNot Nothing Then
				If Not overlayManager.Contains(m_Overlay) Then
					overlayManager.Add(m_Overlay)
				End If 
                DirectCast(m_Overlay, IAgStkGraphicsOverlay).Position = New Object() _
                    { _
                        mouseX, mouseY, _
                        AgEStkGraphicsScreenOverlayUnit.eStkGraphicsScreenOverlayUnitPixels, _
                        AgEStkGraphicsScreenOverlayUnit.eStkGraphicsScreenOverlayUnitPixels _
                    }

                '#Region "CodeSnippet"
                Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager
                '
                ' Get a collection of picked objects in a 100 by 100 rectangular region.
                ' The collection is sorted with the closest object at index zero.
                '
                Dim newModels As New List(Of IAgStkGraphicsModelPrimitive)()
                Dim collection As IAgStkGraphicsPickResultCollection = scene.PickRectangular(attachID("$PickX$The X position to pick at$", mouseX) - attachID("$halfRegionSize$Half of the region size to pick within$", 50), attachID("$PickY$The Y position to pick at$", mouseY) + attachID("$halfRegionSize$Half of the region size to pick within$", 50), attachID("$PickX$The X position to pick at$", mouseX) + attachID("$halfRegionSize$Half of the region size to pick within$", 50), attachID("$PickY$The Y position to pick at$", mouseY) - attachID("$halfRegionSize$Half of the region size to pick within$", 50))
                For Each pickResult As IAgStkGraphicsPickResult In collection
                    Dim objects As IAgStkGraphicsObjectCollection = pickResult.Objects
                    Dim composite As IAgStkGraphicsCompositePrimitive = TryCast(objects(0), IAgStkGraphicsCompositePrimitive)

                    '
                    ' Was a model in our composite picked?
                    '
                    If composite Is attachID("$desiredPrimitive$The primitive to apply the pick action to$", m_Models) Then
                        Dim model As IAgStkGraphicsModelPrimitive = TryCast(objects(1), IAgStkGraphicsModelPrimitive)

                        '
                        ' Selected Model
                        '
                        DirectCast(model, IAgStkGraphicsPrimitive).Color = attachID("$pickedColor$The System.Drawing.Color to change the primitive to when it's picked$", Color.Cyan)
                        newModels.Add(model)
                    End If
                Next
                ' 
                ' Reset color of models that were previous selected but were not in this pick. 
                '
                For Each selectedModel As IAgStkGraphicsModelPrimitive In SelectedModels
                    If Not newModels.Contains(selectedModel) Then
                        DirectCast(selectedModel, IAgStkGraphicsPrimitive).Color = attachID("$notPickedColor$The System.Drawing.Color to change the primitive to when it's not picked$", Color.Red)
                    End If
                Next
                SelectedModels = newModels


					'#End Region
				manager.Render()
			End If
        End Sub

        Public Overrides Sub Execute(ByVal scene As IAgStkGraphicsScene, ByVal root As AgStkObjectRoot)
            ExecuteSnippet(scene, root, Nothing)
        End Sub

		' Name        
		' Description 
		' Category    
		' References  
		' Namespaces  
		' EID         
        <AGI.CodeSnippets.CodeSnippet( _
            "ChangeModelColorInARegion", _
            "Change model colors within a rectangular region", _
            "Graphics | Picking", _
            "System", _
            "System", _
            "AgSTKGraphicsLib~IAgStkGraphicsPickResult" _
            )> _
        Public Sub ExecuteSnippet(<AGI.CodeSnippets.CodeSnippet.Parameter("Scene", "Current Scene")> ByVal scene As IAgStkGraphicsScene, <AGI.CodeSnippets.CodeSnippet.Parameter("Root", "STK Object Model root")> ByVal root As AgStkObjectRoot, <AGI.CodeSnippets.CodeSnippet.Parameter("SelectedModels", "The previously selected models")> ByVal SelectedModels As List(Of IAgStkGraphicsModelPrimitive))
            Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager

            ' Create a screen overlay to visualize the 100 by 100 picking region.
            m_Overlay = manager.Initializers.ScreenOverlay.Initialize(0, 0, 100, 100)
            DirectCast(m_Overlay, IAgStkGraphicsOverlay).PinningOrigin = AgEStkGraphicsScreenOverlayPinningOrigin.eStkGraphicsScreenOverlayPinningOriginCenter
            DirectCast(m_Overlay, IAgStkGraphicsOverlay).Origin = AgEStkGraphicsScreenOverlayOrigin.eStkGraphicsScreenOverlayOriginTopLeft
            DirectCast(m_Overlay, IAgStkGraphicsOverlay).Translucency = 0.9F
            DirectCast(m_Overlay, IAgStkGraphicsOverlay).BorderSize = 2

            Dim r As New Random()

            Dim models As IAgStkGraphicsCompositePrimitive = manager.Initializers.CompositePrimitive.Initialize()
            m_SelectedModels = New List(Of IAgStkGraphicsModelPrimitive)()

            For i As Integer = 0 To 24
                Dim position As Array = New Object() {35 + r.NextDouble(), -(82 + r.NextDouble()), 0.0}

                models.Add(CreateModel(position, root))
            Next

            manager.Primitives.Add(DirectCast(models, IAgStkGraphicsPrimitive))

            OverlayHelper.AddTextBox("Move the rectangular box over models to change their color." & vbCrLf & vbCr & vbLf & _
                                     "This technique, ""roll over"" picking with a rectangular region," & vbCrLf & _
                                     "is implemented by calling Scene.PickRectangular in the 3D " & vbCrLf & _
                                     "window's MouseMove event to determine which primitives are " & vbCrLf & _
                                     "under the rectangular region associated with the mouse.", manager)

            m_Models = DirectCast(models, IAgStkGraphicsPrimitive)
        End Sub

		Private Shared Function CreateModel(position As Array, root As AgStkObjectRoot) As IAgStkGraphicsPrimitive
			Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager

			Dim origin As IAgPosition = root.ConversionUtility.NewPositionOnEarth()
			origin.AssignPlanetodetic(CDbl(position.GetValue(0)), CDbl(position.GetValue(1)), CDbl(position.GetValue(2)))
			Dim axes As IAgCrdnAxesFixed = CreateAxes(root, "Earth", origin)
			Dim system As IAgCrdnSystem = CreateSystem(root, "Earth", origin, axes)
			Dim result As IAgCrdnAxesFindInAxesResult = root.VgtRoot.WellKnownAxes.Earth.Fixed.FindInAxes(DirectCast(root.CurrentScenario, IAgScenario).Epoch, DirectCast(axes, IAgCrdnAxes))

            Dim model As IAgStkGraphicsModelPrimitive = manager.Initializers.ModelPrimitive.InitializeWithStringUri( _
                New AGI.DataPath(AGI.DataPathRoot.Relative, "Models/facility-colorless.mdl").FullPath)
			model.SetPositionCartographic("Earth", position)
			model.Orientation = result.Orientation
			model.Scale = Math.Pow(10, 2)
            DirectCast(model, IAgStkGraphicsPrimitive).Color = Color.Red

			Return DirectCast(model, IAgStkGraphicsPrimitive)
		End Function

		Public Overrides Sub View(scene As IAgStkGraphicsScene, root As AgStkObjectRoot)
            ViewHelper.ViewBoundingSphere(scene, root, "Earth", m_Models.BoundingSphere, -90, 15)
			scene.Camera.Distance *= 0.7
			'zoom in a bit
			scene.Render()
		End Sub

		Public Overrides Sub Remove(scene As IAgStkGraphicsScene, root As AgStkObjectRoot)
			Dim manager As IAgStkGraphicsSceneManager = DirectCast(root.CurrentScenario, IAgScenario).SceneManager
			Dim overlayManager As IAgStkGraphicsScreenOverlayCollectionBase = DirectCast(manager.ScreenOverlays.Overlays, IAgStkGraphicsScreenOverlayCollectionBase)

			overlayManager.Remove(m_Overlay)
			manager.Primitives.Remove(m_Models)
			OverlayHelper.RemoveTextBox(manager)
			scene.Render()

			m_Models = Nothing
			m_SelectedModels = Nothing
		End Sub

		Private m_Models As IAgStkGraphicsPrimitive
		Private m_SelectedModels As List(Of IAgStkGraphicsModelPrimitive)
		Private m_Overlay As IAgStkGraphicsScreenOverlay
	End Class
End Namespace

File: CustomApplications/CSharp/GraphicsHowTo/Picking/PickRectangularCodeSnippet.cs
using System;
using System.IO;
using System.Collections.Generic;

#region UsingDirectives

using System.Collections.ObjectModel;
using System.Drawing;
using AGI.STKGraphics;
using AGI.STKObjects;
using AGI.STKUtil;
using AGI.STKVgt;

#endregion

namespace GraphicsHowTo.Picking
{
    public class PickRectangularCodeSnippet : CodeSnippet
    {
        public PickRectangularCodeSnippet()
            : base(@"Picking\PickRectangularCodeSnippet.cs")
        {
        }

        public void MouseMove(IAgStkGraphicsScene scene, AgStkObjectRoot root, int mouseX, int mouseY)
        {
            IAgStkGraphicsSceneManager manager2 = ((IAgScenario)root.CurrentScenario).SceneManager;
            IAgStkGraphicsScreenOverlayCollectionBase overlayManager = (IAgStkGraphicsScreenOverlayCollectionBase)manager2.ScreenOverlays.Overlays;
            List<IAgStkGraphicsModelPrimitive> SelectedModels = m_SelectedModels;
            if (m_Models != null)
            {
                if (!overlayManager.Contains(m_Overlay))
                {
                    overlayManager.Add(m_Overlay);
                }
                ((IAgStkGraphicsOverlay)m_Overlay).Position = new object[] { mouseX, mouseY, AgEStkGraphicsScreenOverlayUnit.eStkGraphicsScreenOverlayUnitPixels, AgEStkGraphicsScreenOverlayUnit.eStkGraphicsScreenOverlayUnitPixels };

#region CodeSnippet
                IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;
                //
                // Get a collection of picked objects in a 100 by 100 rectangular region.
                // The collection is sorted with the closest object at index zero.
                //
                List<IAgStkGraphicsModelPrimitive> newModels = new List<IAgStkGraphicsModelPrimitive>();
                IAgStkGraphicsPickResultCollection collection = scene.PickRectangular(/*$PickX$The X position to pick at$*/mouseX - /*$halfRegionSize$Half of the region size to pick within$*/50, /*$PickY$The Y position to pick at$*/mouseY + /*$halfRegionSize$Half of the region size to pick within$*/50, /*$PickX$The X position to pick at$*/mouseX + /*$halfRegionSize$Half of the region size to pick within$*/50, /*$PickY$The Y position to pick at$*/mouseY - /*$halfRegionSize$Half of the region size to pick within$*/50);
                foreach (IAgStkGraphicsPickResult pickResult in collection)
                {
                    IAgStkGraphicsObjectCollection objects = pickResult.Objects;
                    IAgStkGraphicsCompositePrimitive composite = objects[0] as IAgStkGraphicsCompositePrimitive;

                    //
                    // Was a model in our composite picked?
                    //
                    if (composite == /*$desiredPrimitive$The primitive to apply the pick action to$*/m_Models)
                    {
                        IAgStkGraphicsModelPrimitive model = objects[1] as IAgStkGraphicsModelPrimitive;

                        //
                        // Selected Model
                        //
                        ((IAgStkGraphicsPrimitive)model).Color = /*$pickedColor$The System.Drawing.Color to change the primitive to when it's picked$*/Color.Cyan;
                        newModels.Add(model);
                    }
                }
                // 
                // Reset color of models that were previous selected but were not in this pick. 
                //
                foreach (IAgStkGraphicsModelPrimitive selectedModel in SelectedModels)
                {
                    if (!newModels.Contains(selectedModel))
                    {
                        ((IAgStkGraphicsPrimitive)selectedModel).Color = /*$notPickedColor$The System.Drawing.Color to change the primitive to when it's not picked$*/Color.Red;
                    }
                }
                SelectedModels = newModels;

                manager.Render();

#endregion
            }
        }

        public override void Execute(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            Execute(scene, root, null);
        }

        [AGI.CodeSnippets.CodeSnippet(
            /* Name        */ "ChangeModelColorInARegion",
            /* Description */ "Change model colors within a rectangular region",
            /* Category    */ "Graphics | Picking",
            /* References  */ "System",
            /* Namespaces  */ "System",
            /* EID         */ "AgSTKGraphicsLib~IAgStkGraphicsPickResult"
            )]
        public void Execute([AGI.CodeSnippets.CodeSnippet.Parameter("Scene", "Current Scene")] IAgStkGraphicsScene scene, [AGI.CodeSnippets.CodeSnippet.Parameter("Root", "STK Object Model root")] AgStkObjectRoot root, [AGI.CodeSnippets.CodeSnippet.Parameter("SelectedModels", "The previously selected models")] List<IAgStkGraphicsModelPrimitive> SelectedModels)
        {
            IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;

            // Create a screen overlay to visualize the 100 by 100 picking region.
            m_Overlay = manager.Initializers.ScreenOverlay.Initialize(0, 0, 100, 100);
            ((IAgStkGraphicsOverlay)m_Overlay).PinningOrigin = AgEStkGraphicsScreenOverlayPinningOrigin.eStkGraphicsScreenOverlayPinningOriginCenter;
            ((IAgStkGraphicsOverlay)m_Overlay).Origin = AgEStkGraphicsScreenOverlayOrigin.eStkGraphicsScreenOverlayOriginTopLeft;
            ((IAgStkGraphicsOverlay)m_Overlay).Translucency = .9f;
            ((IAgStkGraphicsOverlay)m_Overlay).BorderSize = 2;

            Random r = new Random();

            IAgStkGraphicsCompositePrimitive models = manager.Initializers.CompositePrimitive.Initialize();
            m_SelectedModels = new List<IAgStkGraphicsModelPrimitive>();

            for (int i = 0; i < 25; ++i)
            {
                Array position = new object[3] {
                    35 + r.NextDouble(),
                    -(82 + r.NextDouble()), 0.0};

                models.Add(CreateModel(position, root));
            }

            manager.Primitives.Add((IAgStkGraphicsPrimitive)models);

            OverlayHelper.AddTextBox(
@"Move the rectangular box over models to change their color.

This technique, “roll over” picking with a rectangular region,
is implemented by calling Scene.PickRectangular in the 3D 
window's MouseMove event to determine which primitives are 
under the rectangular region associated with the mouse.", manager);

            m_Models = (IAgStkGraphicsPrimitive)models;
        }

        private static IAgStkGraphicsPrimitive CreateModel(Array position, AgStkObjectRoot root)
        {
            IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;

            IAgPosition origin = root.ConversionUtility.NewPositionOnEarth();
            origin.AssignPlanetodetic((double)position.GetValue(0), (double)position.GetValue(1), (double)position.GetValue(2));
            IAgCrdnAxesFixed axes = CreateAxes(root, "Earth", origin);
            IAgCrdnSystem system = CreateSystem(root, "Earth", origin, axes);
            IAgCrdnAxesFindInAxesResult result = root.VgtRoot.WellKnownAxes.Earth.Fixed.FindInAxes(((IAgScenario)root.CurrentScenario).Epoch, ((IAgCrdnAxes)axes));

            IAgStkGraphicsModelPrimitive model = manager.Initializers.ModelPrimitive.InitializeWithStringUri(
                new AGI.DataPath(AGI.DataPathRoot.Relative, "Models/facility-colorless.mdl").FullPath);
            model.SetPositionCartographic("Earth", ref position);
            model.Orientation = result.Orientation;
            model.Scale = Math.Pow(10, 2);
            ((IAgStkGraphicsPrimitive)model).Color = Color.Red;

            return (IAgStkGraphicsPrimitive)model;
        }

        public override void View(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            ViewHelper.ViewBoundingSphere(scene, root, "Earth",
                                               m_Models.BoundingSphere,
                                               -90,
                                               15);
            scene.Camera.Distance *= 0.7; //zoom in a bit
            scene.Render();
        }

        public override void Remove(IAgStkGraphicsScene scene, AgStkObjectRoot root)
        {
            IAgStkGraphicsSceneManager manager = ((IAgScenario)root.CurrentScenario).SceneManager;
            IAgStkGraphicsScreenOverlayCollectionBase overlayManager = (IAgStkGraphicsScreenOverlayCollectionBase)manager.ScreenOverlays.Overlays;
            
            overlayManager.Remove(m_Overlay);
            manager.Primitives.Remove(m_Models);
            OverlayHelper.RemoveTextBox(manager);
            scene.Render();

            m_Models = null;
            m_SelectedModels = null;
        }

        private IAgStkGraphicsPrimitive m_Models;
        private List<IAgStkGraphicsModelPrimitive> m_SelectedModels;
        private IAgStkGraphicsScreenOverlay m_Overlay;
    }
}
